applications/CompressiblePotentialFlowApplication/custom_processes/define_3d_wake_process.cpp

```diff
@@ -168,193 +201,232 @@ void Define3DWakeProcess::InitializeWakeSubModelpart() const
-            }
-            else if(r_geometry[1].GetValue(WING_TIP) != 0){
-                DecreaseWakeWidthAtTheWingTips(coordinates2, coordinates1);
-            }
-        }
+        // To ensure that local_span has the same direction as mSpanDirection
+        const auto& test_start_point = r_geometry[0].Coordinates();
+        const auto& test_end_point   = r_geometry[1].Coordinates();
+        const auto& test_local_span_direction = test_end_point - test_start_point;
+        const double test_projection = inner_prod(test_local_span_direction, mSpanDirection);
 
-        coordinates3 = coordinates1 + mSheddedWakeElementSize * mWakeDirection;
-        coordinates4 = coordinates2 + mSheddedWakeElementSize * mWakeDirection;
+        if (!(test_projection > 0.0))
+        {
+            start_point = r_geometry[1].Coordinates();
+            end_point   = r_geometry[0].Coordinates();
+        }else
+        {
+            start_point = r_geometry[0].Coordinates();
+            end_point   = r_geometry[1].Coordinates();
+        }
 
-        CreateWakeSurfaceNodesAndElements(node_index, coordinates1, coordinates2, coordinates3,
-                                   coordinates4, element_index, pElemProp);
+        // Check if any node is on the wing root
+        if (r_geometry[0].GetValue(WING_ROOT) != 0) {
+            auto local_vector = end_point - start_point;
+            start_point = start_point - (local_vector*mShedProjectionRootEdge/norm_2(local_vector));
+        }
 
-        for (unsigned int j = 0; j < number_of_elements_in_wake_direction; j++){
-            coordinates1 = coordinates3;
-            coordinates2 = coordinates4;
-            coordinates3 = coordinates1 + mSheddedWakeElementSize * mWakeDirection;
-            coordinates4 = coordinates2 + mSheddedWakeElementSize * mWakeDirection;
+        // Create the wake surface
+        double current_element_size = initial_element_size;
+        double accumulated_distance = 0.0;
 
-            CreateWakeSurfaceNodesAndElements(node_index, coordinates1, coordinates2, coordinates3,
-                                   coordinates4, element_index, pElemProp);
-        }
-    }
-}
+        array_1d<double,3> next_start = start_point + current_element_size * mWakeDirection;
+        array_1d<double,3> next_end   = end_point   + current_element_size * mWakeDirection;
 
+        CreateWakeSurfaceNodesAndElements(*mrWakeModelPart, node_index, start_point, end_point, next_start, next_end, element_index, pElemProp);
 
-// This function decreases the wake width, avoiding cutting elements outside of the wake.
-void Define3DWakeProcess::DecreaseWakeWidthAtTheWingTips(array_1d<double, 3>& rPoint1,
-                                                         const array_1d<double, 3>& rPoint2) const
-{
-    const auto& te_edge = rPoint1 - rPoint2;
-    const double projection = inner_prod(te_edge, mSpanDirection);
+        while (accumulated_distance + current_element_size <= max_distance) {
```

Comment:
Why deliberately tend to avoid `while` loops. I'd put a `for` with a maximum number of iterations. If not reached we can throw a warning or even an error.